### PR TITLE
Fix cross-area drag for containers and layouts

### LINKED ISSUE #403: Fix cross-area drag for containers and layouts
### Affected Version of Concrete CMS

9.x (verified against 9.5.x; the affected code paths are unchanged on all active branches)

### Description

You can't drag an existing container block into a different area via drag-and-drop. The drag visibly starts (the title-notch handle responds), but no drop target ever appears for any area other than the container's current area, so the drop never lands anywhere. This includes:

- Dragging a container into another top-level area on the page
- Dragging a container into another container's sub-area
- Dragging a layout block between areas (same code paths, same symptoms)

Three workflows exist for placing a container inside another container's sub-area:

1. **Add Content panel → drag a new container in.** Works.
2. **Area menu → Add Block → select a container.** Works.
3. **Drag an existing container on the page into the sub-area.** Broken — no drop target ever appears.

The sub-area's `data-accepts-block-types` attribute correctly includes `core_container`, and `acceptsBlockType('core_container')` returns `true` at runtime. The data layer is fine. The bug is in the JavaScript drag machinery, which has two separate filters that both restrict container/layout drag to the block's current area only.

Additionally, once those two filters are fixed and cross-area drag succeeds, a third bug surfaces: the moved container's title-notch (the drag/edit/delete handle) is destroyed during the move and never recreated, leaving the nested container with no way to be selected, edited, or re-dragged until the page is reloaded. It seems like this third bug has been present for a long time but never surfaced because the first two bugs made cross-area container drag impossible.

### How to reproduce

Using a fresh Concrete 9.5.x install with the `atomik_full` starting point:

1. Create a new page. Add a **Light Stripe** container (`light_stripe`) and a **Four Column** container (`grid_four`) to the same area.
2. Enter edit mode. Try to drag the Four Column container's title-notch handle into the Light Stripe container's sub-area (Cell 1 or Body).
3. Observe: the drag starts (cursor tracks the handle), but no green drop-target indicator appears anywhere on the page. Releasing the mouse button does nothing.

For comparison, use the Add Content panel (workflow 1 above) to drag a *new* container into Light Stripe's sub-area — the drop target appears immediately and the add succeeds. The sub-area legitimately accepts containers; only the move-existing-container drag path is broken.

#### Bug 1 — `assets/cms/js/edit-mode/containerblock.js` (and `layout.js`)

`Concrete.Block.prototype.setupAreaDragPayloads` iterates all areas and marks each one that accepts the block's type with the `.ccm-area-accepts-block-drag-payload` class:

```js
// Concrete.Block.prototype (base class)
setupAreaDragPayloads: function setupAreaDragPayloads(areas) {
    var my = this
    _(areas).each(function (area) {
        if (area.acceptsBlockType(my.getHandle())) {
            area.getElem().addClass('ccm-area-accepts-block-drag-payload')
        }
    })
}
```

`Concrete.ContainerBlock.prototype` overrides this to ignore the `areas` parameter entirely and only mark the container's current area:

```js
// Concrete.ContainerBlock.prototype (override in containerblock.js)
setupAreaDragPayloads: function setupAreaDragPayloads(areas) {
    var my = this
    my.getArea().getElem().addClass('ccm-area-accepts-block-drag-payload')
}
```

`Concrete.Layout.prototype` in `layout.js` has the identical override.

#### Bug 2 — `assets/cms/js/edit-mode/editmode.js`

Even with Bug 1 patched, the contender selection during mousemove uses a separate filter. In the `EditModeBlockDrag` event handler:

```js
my.bindEvent('EditModeBlockDrag', _.throttle(function (event, data) {
    if (my.getDragging()) {
        var contenders, block = data.block, pep = data.pep, areas = my.getAreas()

        // Bug 2: this line replaces the area list with [block's current area] only
        if (block instanceof Concrete.Layout || block instanceof Concrete.ContainerBlock) {
            areas = [_(areas).find(function (area) { return block.getArea() === area })]
        }

        contenders = _.flatten(_(areas).map(function (area) {
            return area.contendingDragAreas(pep, block)
        }), true)
        // ...
    }
}))
```

When the dragged block is a Layout or ContainerBlock, the area iteration gets replaced with only the block's current area. `contendingDragAreas` is called on just one area, regardless of how many others accept the block. It seems like a second layer of the same cross-area restriction as Bug 1.

#### Bug 3 — `assets/cms/js/edit-mode/block.js`

Once Bugs 1 and 2 are patched and cross-area drag succeeds, the moved container's title-notch disappears from the DOM.

The DOM structure of a container block in edit mode is:

```html
<div class="ccm-area-block-list">
  <div data-container="block">                          <!-- getContainer() returns this -->
    <div class="ccm-edit-mode-title-notch-wrapper">    <!-- the drag handle / menu -->
      <ul class="ccm-edit-mode-title-notch ...">...</ul>
    </div>
    <div class="ccm-block-edit-container ...">         <!-- getElem() returns this -->
      ...the actual block markup...
    </div>
  </div>
</div>
```

The notch wrapper and the block element are siblings inside `<div data-container="block">`. `Concrete.Block.prototype.move` does:

```js
my.getArea().removeBlock(my)
my.getContainer().remove()     // removes the entire wrapper INCLUDING the notch
if (my.getWraps()) {
    // wrapping blocks: build a fresh wrapper
} else {
    myElem = my.getElem()      // non-wrapping blocks: just the bare block element
}
// ...insert myElem at the target...
```

Container blocks have `data-block-type-wraps="0"` (`getWraps()` returns false), so only the bare block element is inserted at the target. The notch wrapper is destroyed.

It seems like this bug has been present for a long time but was never noticed because Bugs 1 and 2 made cross-area container drag impossible — the move handler would only have run for within-area sortable operations where the source and target share the same parent.

#### Bug 4 — `assets/cms/js/edit-mode/containerblock.js` (and `layout.js`)

Once Bugs 1–3 are patched and a cross-area move succeeds, the moved container loses its draggability until the page is reloaded. The drag handle (move icon on the notch) stops responding to mousedown.

The ConcreteMenu click-proxy is a global overlay (`#ccm-menu-click-proxy`, `z-index: 500`, `position: absolute`) that gets positioned over whichever notch the cursor is near. After a move + `scanBlocks()`, the proxy ends up parked over the moved container's notch because the cursor is still in that area. The proxy intercepts mousedown events before they reach the pep-bound drag handle `<li>` underneath.

The drag handle `<li>` sits inside the notch `<ul>` with no explicit positioning or z-index, so it paints below the proxy. On a fresh page load the proxy starts at 0×0 (invisible), which is why the first drag works — the user reaches the move icon before the proxy is positioned there.

Additionally, `ContainerBlock.bindNotchMenu()` creates a `new ConcreteMenu()` without storing a reference to it. When `scanBlocks()` destroys and recreates block instances, the old ConcreteMenu's `mousemove.concreteMenu` handler on the notch is never unbound. These handlers accumulate across drag cycles.

### Why the restriction may exist

I think there are two reasons cross-area drag was disabled for containers and layouts:

1. **Self-nesting prevention.** Dragging container A into one of A's own sub-cells would create infinite recursion. Disabling cross-area drag eliminates that possibility entirely.

2. **Sub-area handle integrity.** Container sub-area handles are prefixed with the parent area handle (`<parentArea> : <instanceID> : <cellName>`). Moving a container to a different area changes the prefix, but `Page::processArrangement()` only updates the container block's own `arHandle` — it doesn't touch the blocks inside the container's sub-areas. Those blocks keep their old handles and stop rendering. Disabling cross-area drag avoids this data-integrity problem.

Both are real concerns, but disabling all cross-area drag is broader than necessary. The self-nesting case is handled by a targeted DOM-descendant check (`draggedBlockElem.contains(areaElem)`), and the sub-area handle problem is addressed by a companion PR that adds handle migration to `processArrangement` (see "Companion fix" below).

### Possible Solution

I have a PR with four changes across `block.js`, `containerblock.js`, `layout.js`, and `editmode.js`:

#### 1. Move the descendant check into the base class (`block.js`)

Delete the `setupAreaDragPayloads` overrides from `containerblock.js` and `layout.js` entirely. Add a descendant check to the base `Block.prototype.setupAreaDragPayloads` instead — the `Element.contains()` call is a no-op for regular blocks (they have no sub-areas in their DOM), so there's no behavioral change for anything except containers and layouts:

```js
setupAreaDragPayloads: function setupAreaDragPayloads(areas) {
    var ownElem = this.getElem()[0]
    _(areas).map((area) => {
        if (ownElem.contains(area.getElem()[0])) return
        if (area.acceptsBlockType(this.getHandle())) {
            area.getElem().addClass('ccm-area-accepts-block-drag-payload')
        }
    })
},
```

#### 2. Replace the `instanceof` filter with a universal descendant check (`editmode.js`)

The `EditModeBlockDrag` handler's `instanceof Concrete.Layout || instanceof Concrete.ContainerBlock` check is replaced with the same `Element.contains()` filter, removing the type-specific special-casing:

```js
var ownElem = block.getElem()[0]
areas = _(areas).filter(function (area) {
    return !ownElem.contains(area.getElem()[0])
})
```

#### 3. Preserve the notch wrapper during cross-area moves (`block.js`)

In `Block.prototype.move`, for non-wrapping blocks (`getWraps() === false`), `detach()` the entire `getContainer()` wrapper instead of calling `remove()` and reinserting only the bare block element. This preserves the sibling notch wrapper (title/drag handle):

```js
var containerWrapper = my.getContainer()
// ...
if (my.getWraps()) {
    containerWrapper.remove()
    myElem = $(targetArea.getBlockTemplate()())
    // ...
} else {
    myElem = containerWrapper.detach()
}
```

#### 4. Ensure the drag handle paints above the ConcreteMenu click-proxy (`containerblock.js`, `layout.js`)

The ConcreteMenu click-proxy overlay (`z-index: 500`) can end up positioned over the container's drag handle after a move, intercepting mousedown events before pep sees them. The drag handle `<li>` now gets `position: relative; z-index: 501` in `bindDrag()` so it always paints above the proxy:

```js
mover.css({ position: 'relative', zIndex: 501 })
```

Also cleans up stale ConcreteMenu event handlers in `bindNotchMenu()` for both ContainerBlock and Layout (`my.getNotch().off('.concreteMenu')`) to prevent handler accumulation across `scanBlocks` cycles.

### Verification

I verified all four changes on a local `atomik_full` install:

1. **Drop targets appear correctly**: dragging a Four Column container shows drop highlights on the top-level Body area and the Light Stripe container's sub-area, but NOT on any of the Four Column container's own four sub-cells. Self-nesting protection is preserved.

2. **End-to-end drag-and-drop succeeds**: dragging the Four Column container into Light Stripe's sub-area completes successfully. The block moves, the title-notch handle survives intact, and the `arrange_blocks` ajax call commits the move to the database.

3. **Round-trip works**: dragging the now-nested Four Column container back out to the top-level Body area also succeeds. Notch survives. No reload needed between drags.

4. **Self-nesting still blocked**: dragging the Four Column container toward one of its own sub-cells correctly shows no drop target.

I have a PR ready against bedrock's `1.7.x` branch.

### Companion fix: sub-area handle migration

Cross-area container moves change the parent area handle prefix that sub-area handles are built from. Without a server-side fix, blocks inside a moved container's sub-areas become orphaned — they're still in the database but no longer render because their `arHandle` values don't match the new sub-area handles.

I've submitted concretecms/concretecms#12865 / concretecms/concretecms#12866, which adds sub-area handle migration to `Page::processArrangement()`. **I'd recommend merging the concretecms PR first** — the migration is harmless until cross-area drag is enabled by this bedrock PR, so it's safe to land early. Once both are merged, populated containers can be moved between areas without content loss.

If for some reason this bedrock PR needs to land first, a `hasContent` guard can be added to `setupAreaDragPayloads` and the `EditModeBlockDrag` handler to restrict populated containers to within-area sorting until the migration catches up:

```js
// In setupAreaDragPayloads (block.js) and EditModeBlockDrag (editmode.js):
var hasContent = $(ownElem).find('.ccm-area[data-total-blocks]').filter(function () {
    return parseInt($(this).attr('data-total-blocks') || '0') > 0
}).length > 0
// Then: if (hasContent && area !== currentArea) return
```

The cleaner path is to land the migration first so the guard isn't needed.

### Additional Context

These bugs are likely a factor in why the friction described in [#11442](https://github.com/concretecms/concretecms/issues/11442) has been hard to address. Enabling cross-area drag for existing containers is a more general fix that also solves the related case of "I want to move a section from one part of the page to another" — complementary to the container-swap proposal discussed there.

### PR DESCRIPTION
**Pull request by janscarton**
Fixes #403

Containers and layouts could only be sorted within their current area. Dragging them to a different area silently failed — no drop target ever appeared. Four issues fixed:

1. **Drop target marking** — ContainerBlock and Layout overrode `setupAreaDragPayloads` to mark only the current area. Deleted the overrides and added a descendant-aware `Element.contains()` check to the base `Block` class instead. Self-nesting is still prevented; all other accepting areas are now valid drop targets.

2. **Contender selection** — The `EditModeBlockDrag` handler replaced the area list with just the current area for Layout/ContainerBlock instances. Replaced with the same universal descendant filter, removing the `instanceof` special-casing.

3. **Notch preservation** — `Block.move()` destroyed the sibling notch wrapper during cross-area moves. For non-wrapping blocks, the entire container wrapper is now `detach()`ed and reinserted as a unit.

4. **Drag handle z-index** — The ConcreteMenu click-proxy overlay (z-index 500) could end up covering the drag handle after a move. The drag handle now gets `position: relative; z-index: 501`. Also cleans up stale ConcreteMenu handlers in `bindNotchMenu()`.

**Companion PR:** concretecms/concretecms#12865 adds sub-area handle migration to `processArrangement()` so that blocks inside a moved container's sub-areas don't get orphaned. I'd recommend merging the concretecms PR first — it's dormant until this bedrock PR enables cross-area drag.

See also concretecms/concretecms#11442 for the broader container workflow discussion.

### CODE REVIEW

**Comment by janscarton** [comment]
This should also fix this concrertecms issue: https://github.com/concretecms/concretecms/issues/12353

**Review by aembler — CHANGES_REQUESTED** [review]
Thanks for taking a look at this, it's much appreciated! We'd definitely accept a pull request to fix the container notch disappearing during edit - that would be a nice thing to fix. The container moving between containers is a tougher sell; migrating the area names is something we haven't done historically because it seems very touchy for the amount of help that it might provide. Additionally, containers are _not_ guaranteed to always look great in different areas, so we have historically not allowed you to move containers between areas.

**Comment by janscarton** [comment]
Thanks, and it makes sense to split these up. I'll pull the cross-area drag out and resubmit the edit-mode fix on its own — the part that keeps the notch and its drag handle from breaking when you move a container around. That stands on its own regardless of cross-area, so it's a clean PR.

I do want to say that disallowing dragging of containers into areas feels like a bug. Moving blocks is core to how editing works, and you can already drag containers around — they just stop at the edge of another area with no message saying why. You can also add a brand-new container inside another container's sub-area, then can't get to that same structure by dragging containers you've already placed. Same end state, one path allowed and the other not.

That said, I get why it's not a mid-cycle fix. I went deep on the handle migration on the concretecms side (#12866) and it's definitely touchy — too many subsystems key on the area handle to rename it safely in a hurry. I'd just track it as a bug rather than intended behavior. Happy to help, though it really wants someone with deeper system knowledge steering it.

**Comment by janscarton** [comment]
Opened the notch + move-control fix on its own as #408 — just that piece, no cross-area drag.

**Comment by janscarton** [comment]
Closing this out. The notch + move-control fix landed separately in #408, and I'm dropping the cross-area drag piece per your review — I get that area-name migration is touchy and that containers aren't guaranteed to look right in other areas. If you ever want to explore cross-area dragging down the road, I'm happy to pick it back up. Just let me know.
